x/ccv/consumer/keeper/relay.go

```diff
@@ -153,55 +183,67 @@ func (k Keeper) SendSlashPacket(ctx sdk.Context, validator abci.Validator, valse
-		k.SetOutstandingDowntime(ctx, consAddr)
+	// send packet over IBC channel
+	err = k.channelKeeper.SendPacket(ctx, channelCap, packet)
+	if clienttypes.ErrClientNotActive.Is(err) {
+		// IBC client expired:
+		// append the Slash packet data to pending data packets
+		// to be sent once the client is upgraded
+		k.AppendPendingDataPacket(ctx, types.DataPacket{
+			Type: types.SlashPacket,
+			Data: packetData.GetBytes(),
+		})
+	} else if err != nil {
+		// something went wrong when sending the packet
+		panic(fmt.Errorf("packet could not be sent over IBC: %w", err))
 	}
 }
 
-// SendPendingSlashRequests iterates over the stored pending slash requests in reverse order
-// and sends the embedded slash packets to the provider chain
-func (k Keeper) SendPendingSlashRequests(ctx sdk.Context) {
-	channelID, ok := k.GetProviderChannel(ctx)
-	if !ok {
-		panic(fmt.Errorf("%s: CCV channel not set", channeltypes.ErrChannelNotFound))
+// SendPendingDataPackets sends the stored pending data packet to the provider chain
+func (k Keeper) SendPendingDataPackets(ctx sdk.Context, channelID string) {
+	dataPackets, found := k.GetPendingDataPackets(ctx)
+	if !found {
+		// This method is a no-op if there are no pending data packets
+		return
 	}
+	for i, dp := range dataPackets.GetList() {
+		// prepare to send the packetData to the consumer
+		packet, channelCap, err := utils.PrepareIBCPacketSend(
+			ctx,
+			k.scopedKeeper,
+			k.channelKeeper,
+			channelID,          // source channel id
+			ccv.ConsumerPortID, // source port id
+			dp.Data,
+			k.GetCCVTimeoutPeriod(ctx),
+		)
+		if err != nil {
+			// something went wrong when preparing the packet
+			panic(fmt.Errorf("packet could not be prepared for IBC send: %w", err))
+		}
 
-	// iterate over pending slash requests in reverse order
-	requests := k.GetPendingSlashRequests(ctx).Requests
-	for i := len(requests) - 1; i >= 0; i-- {
-		slashReq := requests[i]
-
-		// send the emebdded slash packet to the CCV channel
-		// if the outstanding downtime flag is false for the validator
-		downtime := slashReq.Infraction == stakingtypes.Downtime
-		if !downtime || !k.OutstandingDowntime(ctx, sdk.ConsAddress(slashReq.Packet.Validator.Address)) {
-			// send packet over IBC
-			_, err := utils.SendIBCPacket(
-				ctx,
-				k.scopedKeeper,
-				k.channelKeeper,
-				channelID,          // source channel id
-				ccv.ConsumerPortID, // source port id
-				slashReq.Packet.GetBytes(),
-				k.GetCCVTimeoutPeriod(ctx),
-			)
-			if err != nil {
-				panic(err)
-			}
-
-			// set validator outstanding downtime flag to true
-			if downtime {
-				k.SetOutstandingDowntime(ctx, sdk.ConsAddress(slashReq.Packet.Validator.Address))
+		// send packet over IBC channel
+		err = k.channelKeeper.SendPacket(ctx, channelCap, packet)
+		if clienttypes.ErrClientNotActive.Is(err) {
+			// IBC client expired:
+			if i != 0 {
+				// this should never happen
+				panic(fmt.Errorf("client expired while sending pending packets: %w", err))
```

Comment:
This is an edgecase also handled in #448.

I'm wondering if this should invoke a panic?

We could simply delete packets that were sent so we don't resend them when a new client is established?

Reply:
This should never happen. The client cannot expired in the middle of a block as the time doesn't change. Also, it cannot expire in the middle of a loop as the SDK is sequential. I guess we could remove this check to save space :)